numpy/core/src/multiarray/ctors.c

```diff
@@ -819,18 +826,22 @@ PyArray_NewFromDescr_int(
          * which could also be sub-fields of a VOID array
          */
         if (zeroed || PyDataType_FLAGCHK(descr, NPY_NEEDS_INIT)) {
-            data = npy_alloc_cache_zero(nbytes);
+            data = PyDataMem_UserNEW_ZEROED(nbytes, 1,
+                                            &fa->mem_handler->allocator);
         }
         else {
-            data = npy_alloc_cache(nbytes);
+            data = PyDataMem_UserNEW(nbytes, &fa->mem_handler->allocator);
         }
         if (data == NULL) {
             raise_memory_error(fa->nd, fa->dimensions, descr);
             goto fail;
         }
+
         fa->flags |= NPY_ARRAY_OWNDATA;
     }
     else {
+        /* The handlers should never be called in this case */
+        fa->mem_handler = NULL;
```

Comment:
The `mem_handler` field will always be `NULL` when `NPY_OWN_DATA` flag is False, and not `NULL` when `NPY_OWN_DATA` flag is True. That way the allocation strategy of the `base` field will be found in `PyDataMem_GetHandler(obj)`. I checked all the places `NPY_OWN_DATA` is used in `numpy/core/src/multiarray/*`. The only exceptions to this are in the `void` scalar allocations, where `mem_handler` is not used at all.

Reply:
It's important to set this to `NULL`, since the array doesn't own it's own data. `PyDataMem_GetHandler` is currently the "right" API to get the array's true handler (if possible). No SIGSEGV will occur, since NumPy code doesn't use the `mem_handler` at all for arrays that don't own their own data. In cases where internal allocations may happen the current allocator is preferred.

Reply:
If the user sets `NPY_ARRAY_OWNDATA` manually, I guess there should be a way to also set the data allocator?

Reply:
I don't understand.  Where do we store the default allocator when `NPY_ARRAY_OWNDATA` is False?

Reply:
> The mem_handler field will always be NULL when NPY_OWN_DATA flag is False, and not NULL when NPY_OWN_DATA flag is True. That way the allocation strategy of the base field will be found in PyDataMem_GetHandler(obj).

What you are describing here is absolutely right and actually what I tried to explain in my first reply (I forgot to refresh the page and I missed your comment).

In my second reply, I comment on the concern that @seberg raises regarding "what if the user manually sets NPY_ARRAY_OWNDATA afterwards"? I think to handle that case, we should consider storing the current allocator there too (but not using it, except for that corner case).

Reply:
How can the user manually set `NPY_ARRAY_OWNDATA` outside of just setting random values of the `PyArrayObject`?

Reply:
I have no clue, but isn't it what the following [comment](https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/ctors.c#L834-L838) implies?

Reply:
I changed that comment.

Reply:
> How can the user manually set NPY_ARRAY_OWNDATA outside of just setting random values of the PyArrayObject?

Yes, but if for some reason you allocated some data ahead of time and want to transfer ownership to NumPy that may well be the most obvious solution.  Someone is bound to do it?  [First thing I found](https://github.com/davidinouye/destructive-deep-learning/blob/632add7a9731347e050d271ceebb24251e1d8e01/ddl/externals/mlpack/_arma_numpy.pyx#L98)

Reply:
Hmm. That looks like
```
output = numpy.PyArray_SimpleNewFromData(2, &dims[0], numpy.NPY_DOUBLE, GetMemory(X));
if (<logic>) {
    PyArray_ENABLEFLAGS(output, numpy.NPY_OWNDATA);
}
```

At the least this should be turned into
```
output = numpy.PyArray_SimpleNewFromData(2, &dims[0], numpy.NPY_DOUBLE, GetMemory(X));
if (<logic>) {
#if C_API_VERSION >= 0x0f
    output->mem_handler = MyMemHandler
#endif
    PyArray_ENABLEFLAGS(output, numpy.NPY_OWNDATA);
}
```

Maybe we should have a `PyArray_TakeDataOwnership(PyArrayObject, PyDataMem_Handler)` API function that will also set the flags?

I guess that means we need to check for a `NULL` value when using the `mem_handler` :(

Reply:
> Maybe we should have a PyArray_TakeDataOwnership(PyArrayObject, PyDataMem_Handler)

I honestly think we probably don't need it, we just have to put pressure on everyone doing this to set a `->base` that actually owns the data instead.  But if anyone complains about that solution, adding that function seems also good.

Reply:
`OWNDATA` flag and a non-`NULL` `mem_handler` now go hand in hand.

Users should consider using a custom allocator. The comment that encouraged users to just pass new data and set the flag is removed. In complex cases, data ownership can still be transferred to an array by changing the `mem_handler` field, too.

I believe that a memory leak warning is accepted, a SIGSEGV (deallocating an array with the wrong handler) is not. On the other hand, always setting the `default_handler`, instead of `NULL`, could provide backwards compatibility (for the sake of these projects), as long as `default_free` on random addresses doesn't mess with the eventhook/tracemalloc logic that has been also introduced here.

Reply:
> I believe that a memory leak warning is accepted, a SIGSEGV (deallocating an array with the wrong handler) is not. 

How would a SIGSEGV happen when it currently does not?  We are talking only about current users and we currently (de-facto) use a bare `free` in these cases and it works just fine, no?  I assume we can force all new users (and slowly all current ones) to do the correct thing and use a BASE which owns the data.

The point is that I am not very comfortable with breaking these style of dirty "hack this data I got from C" projects, because I think it might be too common (e.g. also in tiny projects run in individual labs/companies).  If there is any semi-reasonable solution here, I would much prefer it.  And that could even be setting a custom `evil-user-detected-allocator` which prints out a warning whenever it is used and supports nothing but `free`.

Reply:
Yes, I got your point. Backwards compatibility is important (even for silently accepted hacks).

I agree, a warning followed by a bare `free`/`realloc` should be enough to cover these cases.